cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/compute/GcpInstanceResourceBuilder.java

```diff
@@ -230,26 +232,43 @@ private void assignToExistingInstanceGroup(GcpContext context,
-            LOGGER.info("adding instance {} to group {} in project {}", instance.getName(), group.getName(), projectId);
-            InstanceGroupsAddInstancesRequest request = createAddInstancesRequest(instance, projectId, zone);
-            AddInstances addInstances = compute.instanceGroups().addInstances(projectId,
-                    zone, instanceGroupResources.get(0).getName(), request);
-            try {
-                Operation execute = addInstances.execute();
-                verifyOperation(execute, buildableResource);
-            } catch (GoogleJsonResponseException e) {
-                LOGGER.error("Error in Google response, unable to add instance {} to group {} : {} for {}",
-                        instance.getName(), group.getName(), e.getMessage(), e.getDetails().getMessage());
-                throw new GcpResourceException(checkException(e), resourceType(), buildableResource.get(0).getName());
-            }
-        } else {
+        List<CloudResource> instanceGroups = filterInstanceGroupsByName(context.getGroupResources(group.getName()), group.getName());
+
+        if (instanceGroups.isEmpty()) {
+            LOGGER.info("skipping group assignment {} was not found in project {}", group.getName(), projectId);
+            return;
+        }
+        if (instanceGroups.size() > 1) {
+            LOGGER.warn("Found multiple instance groups to assign, trying to find an active one. Group {}, project {}", group.getName(), projectId);
+        }
+        CloudResource targetInstanceGroup = findExistingGroup(compute, projectId, zone, instanceGroups);
+
+        if (targetInstanceGroup == null) {
             LOGGER.info("skipping group assignment {} doesn't exist in project {}", group.getName(), projectId);
+            return;
         }
 
+        LOGGER.info("adding instance {} to group {} in project {}", instance.getName(), group.getName(), projectId);
+        InstanceGroupsAddInstancesRequest request = createAddInstancesRequest(instance, projectId, zone);
+        AddInstances addInstances = compute.instanceGroups().addInstances(projectId, zone, targetInstanceGroup.getName(), request);
+        try {
+            Operation execute = addInstances.execute();
+            verifyOperation(execute, buildableResource);
+        } catch (GoogleJsonResponseException e) {
+            LOGGER.error("Error in Google response, unable to add instance {} to group {} : {} for {}",
+                    instance.getName(), group.getName(), e.getMessage(), e.getDetails().getMessage());
+            throw new GcpResourceException(checkException(e), resourceType(), buildableResource.get(0).getName());
+        }
+    }
+
+    @Nullable
+    private CloudResource findExistingGroup(Compute compute, String projectId, String zone, List<CloudResource> candidates) throws IOException {
+        for (CloudResource candidate : candidates) {
+            if (doesGcpInstanceGroupExist(compute, projectId, zone, candidate)) {
```

Comment:
Please try to optimize `doesGcpInstanceGroupExist` as it issues `compute.instanceGroups().list(..)` multiple times.

Reply:
done

Reply:
optimized